compiler/aarch64/codegen/BinaryEvaluator.cpp

```diff
@@ -179,6 +179,71 @@ generateMaddOrMsub(TR::Node *node, TR::Node *mulNode, TR::Node *anotherNode, TR:
       }
    }
 
+/**
+ * @brief Generates add (shifted register) or sub (shifted register) instruction if possible
+ *
+ * @param[in]  node:  node
+ * @param[in]    op:  mnemonic for this node
+ * @param[in]    cg:  code generator
+ *
+ * @return register which contains the result of the operation. NULL if the operation cannot be encoded in add or sub shifted register instruction.
+ */
+static TR::Register *
+generateAddOrSubShifted(TR::Node *node, TR::InstOpCode::Mnemonic op, TR::CodeGenerator *cg)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Node *source1Node = NULL;
+   TR::Node *source2Node = NULL;
+   TR::Node *shiftNode = NULL;
+
+   int32_t shiftValue;
+
+   if ((!firstChild->getOpCode().isLoadConst()) &&
+       (secondChild->getReferenceCount() == 1) &&
+       (secondChild->getRegister() == NULL) &&
+       secondChild->getOpCode().isShift() &&
+       (secondChild->getSecondChild()->getOpCodeValue() == TR::iconst))
+      {
+      source1Node = firstChild;
+      source2Node = secondChild->getFirstChild();
+      shiftValue = secondChild->getSecondChild()->getInt();
+      shiftNode = secondChild;
+      }
+   else if (node->getOpCode().isAdd() &&
+             (!secondChild->getOpCode().isLoadConst()) &&
+             (firstChild->getReferenceCount() == 1) &&
+             (firstChild->getRegister() == NULL) &&
+             firstChild->getOpCode().isShift() &&
+             (firstChild->getSecondChild()->getOpCodeValue() == TR::iconst))
+      {
+      source1Node = secondChild;
+      source2Node = firstChild->getFirstChild();
+      shiftValue = firstChild->getSecondChild()->getInt();
+      shiftNode = firstChild;
+      }
+   else
+      {
+      return NULL;
+      }
+
+   const bool is64bit = node->getDataType().isInt64();
+   if ((shiftValue <= 0) || (shiftValue > (is64bit ? 63 : 31)))
+      {
+      return NULL;
+      }
+   TR::Register *treg = cg->allocateRegister();
```

Comment:
Is it possible to check the reference count of source1Node and source2Node here, to reuse s1reg or s2reg as treg?

Reply:
Thanks for the suggestion. Will work on it.

Reply:
force-pushed the updated commit